book-content/chapters/07-mutability.md

```diff
@@ -0,0 +1,1013 @@
+
+##### Option 2: Use the `ReadonlyArray` Type Helper
+
+Alternatively, you could use the `ReadonlyArray` type helper:
+
+```typescript
+function printNames(names: ReadonlyArray<string>) {
+  ...
+}
+```
+
+Regardless of which of these two methods you use, TypeScript will still display `readonly string[]` when hovering over the `names` parameter:
+
+```typescript
+// hovering over `names` shows:
+(parameter) names: readonly string[]
+```
+
+Both work equally well at preventing the array from being modified.
+
+#### Solution 3: An Unsafe Tuple
+
+The best way to prevent unwanted changes to the `Coordinate` tuple is to make it a `readonly` tuple:
+
+```typescript
+type Coordinate = readonly [number, number];
+```
+
+Now, `dangerousFunction` throws a TypeScript error when we try to pass `myHouse` to it:
+
+```ts twoslash
+// @errors: 2345
+type Coordinate = readonly [number, number];
+const myHouse: Coordinate = [0, 0];
+
+// ---cut---
+const dangerousFunction = (arrayOfNumbers: number[]) => {
+  arrayOfNumbers.pop();
+  arrayOfNumbers.pop();
+};
+
+dangerousFunction(myHouse);
+```
+
+We get an error because the function's signature expects a modifiable array of numbers, but `myHouse` is a read-only tuple. TypeScript is protecting us against unwanted changes.
+
+It's a good practice to use `readonly` tuples as much as possible to avoid problems like the one in this exercise.
+
+## Deep Immutability with `as const`
+
+We've seen so far that objects and arrays are mutable in JavaScript. This leads to their properties being inferred _widely_ by TypeScript.
+
+We can get around this by giving the property a type annotation. But it still doesn't infer the literal type of the property.
+
+```typescript
+const albumAttributes: AlbumAttributes = {
+  status: "on-sale",
+};
+
+// hovering over albumAttributes shows:
+const albumAttributes: {
+  status: "new-release" | "on-sale" | "staff-pick";
+};
+```
+
+Instead of `albumAttributes.status` being inferred as `"on-sale"`, it's inferred as `"new-release" | "on-sale" | "staff-pick"`.
+
+One way we could get TypeScript to infer it properly would be to somehow mark the entire object, and all its properties, as immutable. This would tell TypeScript that the object and its properties can't be changed, so it would be free to infer the literal types of the properties.
+
+This is where the `as const` assertion comes in. We can use it to mark an object and all of its properties as constants, meaning that they can't be changed once they are created.
+
+```typescript
+const albumAttributes = {
+  status: "on-sale",
+} as const;
+
+// hovering over albumAttributes shows:
+const albumAttributes: {
+  readonly status: "on-sale";
+};
+```
+
+The `as const` assertion has made the entire object deeply read-only, including all of its properties. This means that `albumAttributes.status` is now inferred as the literal type `"on-sale"`.
```

Comment:
It might be worth noting here that it will only work for "fresh" parts of `as const`ed object: [TS playground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAZiEMC8MDeAoGMBGBDAJwC4YAiUgGgwF8BuDDUSWEHAKxXS3kSupjwQYTaPQwB6cTAACUCAFoApgA8ADouBQlBAiAIZWbAHQIkqTNnzEyYRQHdKNMYZOIjVzgHJ8ALwCWYADmeACEYZ4wGEA)